core/search/phrase_iterator.hpp

```diff
@@ -23,34 +23,187 @@
 
 NS_ROOT
 
-// implementation is optimized for frequency based similarity measures
-// for generic implementation see a03025accd8b84a5f8ecaaba7412fc92a1636be3
-template<typename Conjunction>
-class phrase_iterator final : public doc_iterator_base {
+class fixed_phrase_frequency {
  public:
   typedef std::pair<
     position::ref, // position attribute
     position::value_t // desired offset in the phrase
   > position_t;
   typedef std::vector<position_t> positions_t;
 
+  fixed_phrase_frequency(
+      positions_t&& pos,
+      const order::prepared& ord
+  ) : pos_(std::move(pos)), order_(&ord) {
+    assert(!pos_.empty()); // must not be empty
+    assert(0 == pos_.front().second); // lead offset is always 0
+  }
+
+ protected:
+  // returns frequency of the phrase
+  frequency::value_t phrase_freq() {
+    frequency::value_t freq = 0;
+    bool match;
+
+    position& lead = pos_.front().first;
+    lead.next();
+
+    for (auto end = pos_.end(); !pos_limits::eof(lead.value());) {
+      const position::value_t base_position = lead.value();
+
+      match = true;
+
+      for (auto it = pos_.begin() + 1; it != end; ++it) {
+        position& pos = it->first;
+        const auto term_position = base_position + it->second;
+        if (!pos_limits::valid(term_position)) {
+          return freq;
+        }
+        const auto seeked = pos.seek(term_position);
+
+        if (pos_limits::eof(seeked)) {
+          // exhausted
+          return freq;
+        } else if (seeked != term_position) {
+          // seeked too far from the lead
+          match = false;
+
+          lead.seek(seeked - it->second);
+          break;
+        }
+      }
+
+      if (match) {
+        if (order_->empty()) {
+          return 1;
+        }
+
+        ++freq;
+        lead.next();
+      }
+    }
+
+    return freq;
+  }
+
+ private:
+  positions_t pos_; // list of desired positions along with corresponding attributes
+  const order::prepared* order_;
+}; // fixed_phrase_frequency
+
+class variadic_phrase_frequency {
+ public:
+  typedef std::pair<
+    disjunction_visitor<position_score_iterator_adapter<doc_iterator::ptr>>*,
+    position::value_t // desired offset in the phrase
+  > position_t;
+  typedef std::vector<position_t> positions_t;
+
+  variadic_phrase_frequency(
+      positions_t&& pos,
+      const order::prepared& ord)
+    : pos_(std::move(pos)), order_(&ord) {
+    assert(!pos_.empty()); // must not be empty
+    assert(0 == pos_.front().second); // lead offset is always 0
+    is_iterators_hitched_.resize(pos_.size());
+  }
+
+ protected:
+  // returns frequency of the phrase
+  frequency::value_t phrase_freq() {
+    frequency::value_t freq = 0;
+    position::value_t term_position = pos_limits::eof();
+    uint32_t min_seeked = pos_limits::eof();
+    auto match = false;
+    // reset is_hitched
+    is_iterators_hitched_.assign(is_iterators_hitched_.size(), 0);
+
+    auto innerVisitor = [&term_position, &min_seeked, &match](
```

Comment:
consider creation of members of type `std::function<...>` to avoid frequent and expensive conversions from lambda to `std::function`, as an alternative to `std::function<...>` consider raw function pointer with context, e.g.
`void (*visitor)(const void* ctx, const Adapter&)`